gnome-initial-setup/pages/endless-eula/gis-endless-eula-viewer.c

```diff
@@ -0,0 +1,298 @@
+  const gchar *product_name;
+  const gchar *language;
+  g_autofree gchar *path = NULL;
+
+  driver = GIS_DRIVER (g_application_get_default ());
+  language = gis_driver_get_user_language (driver);
+  locale_variants = g_get_locale_variants (language);
+  product_name = gis_driver_get_product_name (driver);
+
+  path = find_terms_document_for_languages (product_name, (const gchar * const *)locale_variants);
+  if (path == NULL)
+    path = find_terms_document_for_languages ("eos", (const gchar * const *)locale_variants);
+
+  if (path == NULL)
+    {
+      g_critical ("Unable to find terms and conditions PDF on the system");
+      return NULL;
+    }
+
+  return g_file_new_for_path (path);
+}
+
+static gboolean
+gis_endless_eula_viewer_decide_policy_cb (GisEndlessEulaViewer     *self,
+                                          WebKitPolicyDecision     *decision,
+                                          WebKitPolicyDecisionType  decision_type,
+                                          gpointer                  user_data)
+{
+  g_return_val_if_fail (GIS_IS_ENDLESS_EULA_VIEWER (self), FALSE);
+
+  if (decision_type != WEBKIT_POLICY_DECISION_TYPE_NAVIGATION_ACTION)
+    return FALSE;
+
+  WebKitNavigationPolicyDecision *navigation_decision = WEBKIT_NAVIGATION_POLICY_DECISION (decision);
+  WebKitNavigationAction *action = webkit_navigation_policy_decision_get_navigation_action (navigation_decision);
+  WebKitURIRequest *request = webkit_navigation_action_get_request (action);
+  const gchar *uri = webkit_uri_request_get_uri (request);
+
+  if (g_strcmp0 (uri, self->terms_uri) == 0)
+    {
+      webkit_policy_decision_use (decision);
+      return TRUE;
+    }
+
+  webkit_policy_decision_ignore (decision);
+
+  const gchar *title = NULL;
+
+  if (g_str_has_prefix (uri, "file:"))
+    title = _("Terms of Use");
+  else if (g_str_has_prefix (uri, LICENSE_SERVICE_URI))
+    title = _("Third Party Software");
+  else
+    /* Ignore attempts to navigate to remote pages: at this point in Initial
+     * Setup the user has not had a chance to configure Wi-Fi so is almost
+     * certainly offline.
+     */
+    return TRUE;
+
+  show_uri_in_modal (self, uri, title);
+  return TRUE;
+}
+
+static gboolean
+gis_endless_eula_viewer_context_menu_cb (WebKitWebView       *self,
+                                         WebKitContextMenu   *context_menu,
+                                         WebKitHitTestResult *hit_test_result,
+                                         gpointer             user_data)
+{
+  /* Hide default context menu: all useful actions are available in the PDF.js
+   * toolbar.
+   */
+  return TRUE;
+}
+
+static void
+load_terms_view (GisEndlessEulaViewer *self)
+{
+  g_autoptr(GFile) file = NULL;
+
+  file = get_terms_document ();
+  if (file == NULL)
+    return;
+
+  g_assert (self->terms_uri == NULL);
+  self->terms_uri = g_file_get_uri (file);
+  webkit_web_view_load_uri (WEBKIT_WEB_VIEW (self), self->terms_uri);
```

Comment:
Right now we aren't showing anything if the URI fails to load, but this is not a regression from the previous version. However, this document does take a noticeable time to load. Should we add a GtkOverlay with a progress bar here, similar to the one in `show_uri_in_modal`?

Reply:
We could. This page is loaded in the background, while the user is looking at the language & keyboard layout (and possibly live-chooser) pages, but I guess we can do that anyway.

Reply:
I tested this further and you're right, the PDF is actually only loaded once the page is displayed. I think it's fast enough but a progress bar might be nice as a future iteration.